### PR TITLE
[#3] Set up CD with GitHub Actions

### LINKED ISSUE #3: Set up CD with GitHub Actions
## Why

We need a way to build a project Docker image and put this image in the registry.

## Acceptance Criteria

- When code pushed to `develop` branch staging version of container should be build and pushed to AWS ECR
- When code pushed to `main` branch production version of container should be build and pushed to AWS ECR

### PR DESCRIPTION
**Pull request by Nihisil**
Close #3

## What happened 👀

Add GH action workflow to build docker image and deploy application to AWS.

Same workflow supports staging deployment from `develop` branch and production deployment from `main` branch.

## Proof Of Work 📹

You can find test workflow run here, image was built and deployed: https://github.com/Nihisil/nimble-devops-ic-app/actions/runs/4051871868

### CODE REVIEW

File: .github/workflows/build_and_deploy.yml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        uses: aws-actions/amazon-ecr-login@v1.5.3
+
+      - name: Build, tag, and push image to AWS ECR
+        id: build-image
+        env:
+          SHARED_ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+        run: |
+          docker build -t $SHARED_ECR_REGISTRY/$ECR_REPOSITORY:$CI_REF_NAME_SLUG-$CI_SHA_SHORT -t $SHARED_ECR_REGISTRY/$ECR_REPOSITORY:staging .
+          docker push -a $SHARED_ECR_REGISTRY/$ECR_REPOSITORY
+          echo "::set-output name=image::$SHARED_ECR_REGISTRY/$ECR_REPOSITORY:$CI_REF_NAME_SLUG-$CI_SHA_SHORT"
```
Comment (andyduong1920) [inline .github/workflows/build_and_deploy.yml]: The `set output` is deprecating https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/, it's time to use a new method for this @Nihisil ;) 

https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter

Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: Thanks for pointing on it. Usage of `set-output` was removed here: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/b963fec06d585f33d3b2ec693b52944e6f808a4a

With other minor refactorings.

File: .github/workflows/build_and_deploy.yml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    name: Build the Docker image
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    outputs:
+      image: ${{ steps.build-image.outputs.image }}
+
+    steps:
+      - name: Debug
+        run: |
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: I think this step is no longer needed, right? :D

File: .github/workflows/build_and_deploy.yml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+          echo "ECR_REPOSITORY=${{ env.ECR_REPOSITORY }}"
+          echo "REGION=${{ env.REGION }}"
+          echo "env_name=${{ needs.shared_variables.outputs.env_name }}"
+          echo "container_name=${{ needs.shared_variables.outputs.container_name }}"
+          echo "ecs_cluster=${{ needs.shared_variables.outputs.ecs_cluster }}"
+          echo "ecs_service=${{ needs.shared_variables.outputs.ecs_service }}"
+          echo "ecs_task_definition=${{ needs.shared_variables.outputs.ecs_task_definition }}"
+      
+      - name: ENV variables
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: Follow Nimble's naming [convention](https://nimblehq.co/compass/development/code-conventions/github-actions/#convention-5), the name should start with a verb to describe the intention
```suggestion
      - name: Expose useful environment variables
```

File: .github/workflows/build_and_deploy.yml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        uses: FranzDiebold/github-env-vars-action@v2.7.0
+
+      - name: Checkout
+        uses: actions/checkout@v2
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: There is [newer versions](https://github.com/actions/checkout/releases) of this action 
```suggestion
        uses: actions/checkout@v3
```

File: .github/workflows/build_and_deploy.yml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  
+  deploy:
+    needs: [build_image, shared_variables]
+    name: Deploy the built image
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1.6.1
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: how about using the same version as in the `build_image` step
```suggestion
        uses: aws-actions/configure-aws-credentials@v1.7.0
```

File: .github/workflows/build_and_deploy.yml [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1.7.0
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ env.REGION }}
+          mask-aws-account-id: 'no'
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: we had to remove this `mask-aws-account-id` option due to some issue which I don't remember exactly. Maybe @andyduong1920 could shed some light as he faced it.
Found these issues on the action repo:
- https://github.com/aws-actions/configure-aws-credentials/issues/459
- https://github.com/aws-actions/configure-aws-credentials/issues/304

**Review by longnd — COMMENTED** [review]
Should we take this chance to update the build badge here?
https://github.com/Nihisil/nimble-devops-ic-app/blob/78fda7cb734f5c86643d877b849b54f64406d4de/README.md?plain=1#L1

https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/adding-a-workflow-status-badge

File: .github/workflows/build_and_deploy.yml (hunk 2 above)
Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: I was sure I removed it before 🤦‍♂️ . It was removed here: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/6e333979df0e161cc085ad6d9aab4f809208ca9a

File: .github/workflows/build_and_deploy.yml (hunk 3 above)
Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: Thanks. It was updated there: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/99cdd3bc7ba173579f52d4e202f69e41881de318

File: .github/workflows/build_and_deploy.yml (hunk 4 above)
Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: All versions were upgraded here: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/c16ac1f00ce5f796ad404c21d5afadaf45f5e032

File: .github/workflows/build_and_deploy.yml (hunk 5 above)
Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: All versions were upgraded here: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/c16ac1f00ce5f796ad404c21d5afadaf45f5e032

File: .github/workflows/build_and_deploy.yml (hunk 6 above)
Comment (Nihisil) [reply .github/workflows/build_and_deploy.yml]: Thanks for pointing on it. I was sure that this is required attr, but it is not. It was removed here: https://github.com/Nihisil/nimble-devops-ic-app/pull/12/commits/8044c19ea1cb1e44ad70909ee0610b706ec9c20a

Comment (andyduong1920) [reply .github/workflows/build_and_deploy.yml]: We just don't need that actually :D

**Review by andyduong1920 — APPROVED** [review]

**Comment by Nihisil** [comment]
@andyduong1920 @longnd 

I have simplified the setup a bit more. Since we are passing many variables between jobs, we can reduce complexity by combining everything in one job. 

The resulting workflow is not too big, so I think it is fine to keep it as is, and there is no real need to separate it on different jobs as it was done before.

You can check that the run is still working and the app is being built and deployed to ECS: https://github.com/Nihisil/nimble-devops-ic-app/actions/runs/4060976502

File: .github/workflows/build_and_deploy.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      - main
+  workflow_dispatch:
+
+env:
+  ECR_REPOSITORY: alex-ic
+  REGION: ap-southeast-1
+
+jobs:
+  build_and_deploy:
+    name: Build the Docker image
```
Comment (longnd) [inline .github/workflows/build_and_deploy.yml]: The name here is misleading as it indicates there is just the build step but actually the step does more than that
```suggestion
    name: Build the Docker image and deploy to AWS
```
Screenshot 2023-02-01 at 12 17 49

**Review by longnd — APPROVED** [review]
nice work, the workflow is much simpler 👍

**Pull request merged**